### PR TITLE
Update precompile setup for warp

### PR DESCRIPTION
**Pull request by aaronbuchwald**
This PR migrates Coreth to use the latest precompile package structure from Subnet-EVM in preparation for migrating the Warp precompile.

### CODE REVIEW

File: core/vm/contracts.go [hunk 1]
```diff
@@ -202,30 +203,21 @@ func init() {
 	addrsList = append(addrsList, PrecompiledAddressesApricotPhasePre6...)
 	addrsList = append(addrsList, PrecompiledAddressesApricotPhase6...)
 	addrsList = append(addrsList, PrecompiledAddressesBanff...)
+	addrsList = append(addrsList, genesisContractAddr, NativeAssetBalanceAddr, NativeAssetCallAddr) // Add genesis multicoin contract and deprecated native asset precompiles to set of reserved addresses
```
Comment (darioush) [inline core/vm/contracts.go:206]: why is this needed? seems they are included in PrecompiledContractsApricotPhase2

File: core/vm/contracts.go [hunk 2]
```diff
@@ -202,30 +203,21 @@ func init() {
...
-		found := false
-		for _, reservedRange := range precompile.ReservedRanges {
-			if reservedRange.Contains(k) {
-				found = true
-				break
-			}
-		}
-		if !found {
-			panic(fmt.Errorf("address %s used for stateful precompile but not specified in any reserved range", k))
+		if address == constants.BlackholeAddr {
```
Comment (darioush) [inline core/vm/contracts.go:219]: why do we check this here vs in subnet-evm where it's checked here https://github.com/ava-labs/subnet-evm/blob/32514d6c7c6523d08a07eadda9919ef2253232de/precompile/modules/registerer.go#L52-L53

File: params/avalanche_params.go [hunk 3]
```diff
@@ -32,6 +31,9 @@ const (
 	ApricotPhase5TargetGas                uint64 = 15_000_000
 	ApricotPhase5BaseFeeChangeDenominator uint64 = 36
 
+	DynamicFeeExtraDataSize        = 80
+	RollupWindow            uint64 = 10
```
Comment (darioush) [inline params/avalanche_params.go:35]: doesn't seem this is referenced in the code.

File: core/evm.go [hunk 4]
```diff
@@ -63,17 +92,16 @@ func NewEVMBlockContext(header *types.Header, chain ChainContext, author *common
 		baseFee = new(big.Int).Set(header.BaseFee)
 	}
 	return vm.BlockContext{
-		CanTransfer:       CanTransfer,
-		CanTransferMC:     CanTransferMC,
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply core/evm.go:67]: Good catch, will update

File: core/vm/contracts.go (hunk 1 above)
Comment (aaronbuchwald) [reply core/vm/contracts.go:206]: Good catch

File: params/avalanche_params.go (hunk 3 above)
Comment (aaronbuchwald) [reply params/avalanche_params.go:35]: This is migrated in the next PR

File: core/vm/contracts.go (hunk 2 above)
Comment (aaronbuchwald) [reply core/vm/contracts.go:219]: Removing

**Review by darioush — APPROVED** [review]

File: core/txpool/txpool.go [hunk 5]
```diff
@@ -269,10 +269,11 @@ type TxPool struct {
 	signer      types.Signer
 	mu          sync.RWMutex
 
-	istanbul atomic.Bool // Fork indicator whether we are in the istanbul stage.
-	eip2718  atomic.Bool // Fork indicator whether we are using EIP-2718 type transactions.
-	eip1559  atomic.Bool // Fork indicator whether we are using EIP-1559 type transactions.
-	eip3860  atomic.Bool // Fork indicator whether EIP-3860 is activated. (activated in Shanghai Upgrade in Ethereum)
+	rules    atomic.Pointer[params.Rules] // Rules for the currentHead
+	istanbul atomic.Bool                  // Fork indicator whether we are in the istanbul stage.
```
Comment (ceyonur) [inline core/txpool/txpool.go:273]: I don't think we need this anymore

**Review by ceyonur — COMMENTED** [review]
just few nits

File: params/config.go [hunk 6]
```diff
@@ -484,6 +486,19 @@ var (
 	TestRules = TestChainConfig.AvalancheRules(new(big.Int), 0)
 )
 
+// UpgradeConfig includes the following configs that may be specified in upgradeBytes:
+// - Timestamps that enable avalanche network upgrades,
+// - Enabling or disabling precompiles as network upgrades.
+type UpgradeConfig struct {
+	// Config for enabling and disabling precompiles as network upgrades.
+	PrecompileUpgrades []PrecompileUpgrade `json:"precompileUpgrades,omitempty"`
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply params/config.go:494]: Yes it's activated with the DUpgrade in `vm.go`. I thought it was easier to keep Coreth and Subnet-EVM closer in line rather than having a different mechanism.

Additionally, we can't import the Warp precompile in the `params/` package, so it needs to be done externally anyways. Open to suggestion, but after playing around a bit, I thought the best path was to simply share the same code with Subnet-EVM and set the `PrecompileUpgrades` field in `vm.go` here: https://github.com/ava-labs/coreth/blob/migrate-warp-precompile/plugin/evm/vm.go#L470

File: params/precompile_upgrade.go
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply params/precompile_upgrade.go:1]: https://github.com/ava-labs/coreth/pull/390#discussion_r1407914445

File: core/vm/contracts_stateful.go [hunk 7]
```diff
@@ -51,7 +51,7 @@ type nativeAssetBalance struct {
 }
 
 // PackNativeAssetBalanceInput packs the arguments into the required input data for a transaction to be passed into
-// the native asset balance precompile.
+// the native asset balance contract.
 func PackNativeAssetBalanceInput(address common.Address, assetID common.Hash) []byte {
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply core/vm/contracts_stateful.go:55]: Good call

**Review by ceyonur — DISMISSED** [review]

File: params/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ceyonur) [reply params/config.go:494]: I was just in favor of having less code to reduce risk. But I think that's fine.

**Review by darioush — APPROVED** [review]
